Pinta.Effects/Effects/VoronoiDiagramEffect.cs

```diff
@@ -0,0 +1,284 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Linq;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class VoronoiDiagramEffect : BaseEffect
+{
+	// TODO: Icon
+
+	public override bool IsTileable => false;
+
+	public override string Name => Translations.GetString ("Voronoi Diagram");
+
+	public override bool IsConfigurable => true;
+
+	public override string EffectMenuCategory => Translations.GetString ("Render");
+
+	public VoronoiDiagramData Data => (VoronoiDiagramData) EffectData!; // NRT - Set in constructor
+
+	public VoronoiDiagramEffect ()
+	{
+		EffectData = new VoronoiDiagramData ();
+	}
+
+	public override void LaunchConfiguration ()
+		=> EffectHelper.LaunchSimpleEffectDialog (this);
+
+	private sealed record VoronoiSettings (
+		int w,
+		int h,
+		bool showPoints,
+		ImmutableArray<PointI> points,
+		ImmutableArray<ColorBgra> colors,
+		Func<PointI, PointI, double> distanceCalculator);
+
+	private VoronoiSettings CreateSettings (ImageSurface dst, RectangleI roi)
+	{
+		ColorSorting colorSorting = Data.ColorSorting;
+
+		IEnumerable<PointI> basePoints = CreatePoints (roi, Data.PointCount, Data.PointLocationsSeed);
+		ImmutableArray<PointI> points = SortPoints (basePoints, colorSorting).ToImmutableArray ();
+
+		IEnumerable<ColorBgra> baseColors = CreateColors (points.Length, Data.ColorsSeed);
+		IEnumerable<ColorBgra> positionSortedColors = SortColors (baseColors, colorSorting);
+		IEnumerable<ColorBgra> reversedSortingColors = Data.ReverseColorSorting ? positionSortedColors.Reverse () : positionSortedColors;
+
+		return new (
+			w: dst.Width,
+			h: dst.Height,
+			showPoints: Data.ShowPoints,
+			points: points,
+			colors: reversedSortingColors.ToImmutableArray (),
+			distanceCalculator: GetDistanceCalculator (Data.DistanceCalculationMethod)
+		);
+	}
+
+	protected override void Render (ImageSurface src, ImageSurface dst, RectangleI roi)
+	{
+		VoronoiSettings settings = CreateSettings (dst, roi);
+
+		Span<ColorBgra> dst_data = dst.GetPixelData ();
+
+		for (int y = roi.Top; y <= roi.Bottom; y++) {
+			var dst_row = dst_data.Slice (y * settings.w, settings.w);
+			for (int x = roi.Left; x <= roi.Right; x++) {
+				PointI pixelLocation = new (x, y);
+				double shortestDistance = double.MaxValue;
+				int closestIndex = 0;
+				for (var i = 0; i < settings.points.Length; i++) {
+					var point = settings.points[i];
+					double distance = settings.distanceCalculator (point, pixelLocation);
+					if (distance > shortestDistance) continue;
+					shortestDistance = distance;
+					closestIndex = i;
+				}
+				dst_row[x] =
```

Comment:
Only using the closest point can produce jagged edges, so you may want to consider doing some sort of blending / antialiasing

Reply:
Maybe this would require either changing the algorithm used to make the calculations, or applying an edge blur effect?

Reply:
It could probably be done easily with supersampling, e.g. compute the color at multiple sub-pixel positions and blend - currently the sampling is done at only the upper left corner of the pixel

Most pixels don't need this though since there are large solid regions, so this could be skipped if there isn't another voronoi point close enough to be the closest to any sub pixel location

Reply:
Can this be done in a future PR? There's no risk of breaking an API if we add stuff to it.

Reply:
Yeah I think this is fine to add in a future PR